substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/jfr/remote/JfrAutoSessionManager.java

```diff
@@ -0,0 +1,152 @@
+    static JfrAutoSessionManager instance;
+
+    static void out(String msg) {
+        JfrLogger.logInfo("JFR.AutoStart " + msg);
+    }
+
+    JfrAutoSessionManager() throws IOException, ParseException {
+        this(DEFAULT_JFR_CONFIG);
+    }
+
+    JfrAutoSessionManager(String configName) throws IOException, ParseException {
+        out("JfrAutoSessionManager(" + configName + ")");
+        jfrConfiguration = Configuration.getConfiguration(configName);
+    }
+
+    /*
+     *   JFR-TODO if these timers expire after the main program has ended, the process hangs.
+     */
+    void startSession(String fn, long delayMilliseconds, long durationMilliseconds) {
+        if (delayMilliseconds > 0) {
+            out("scheduling JFR start in " + delayMilliseconds + "ms.");
+            new java.util.Timer().schedule(
+                    new java.util.TimerTask() {
+                        @Override
+                        public void run() {
+                            startSession(fn, 0, durationMilliseconds);
+                        }
+                    },
+                    delayMilliseconds
+            );
+        } else {
+            out("starting JFR session(" + fn + ")");
+            filename = fn;
+            recording = new Recording(jfrConfiguration);
+            recording.start();
+            try {
+                recording.setDestination(Paths.get(filename));
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            if (durationMilliseconds > 0) {
+                out("scheduling JFR stop in " + durationMilliseconds + "ms.");
+                new java.util.Timer().schedule(
+                        new java.util.TimerTask() {
+                            @Override
+                            public void run() {
+                                stopSession(true, false);
+                            }
+                        },
+                        durationMilliseconds
+                );
+            }
+        }
+    }
+
+    void stopSession(boolean write, boolean clear) {
+        out("stopSession(" + write + ", " + clear + ") start");
+        try {
+            if (recording != null) {
+                recording.close();
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        out("stopSession(" + write + ", " + clear + ") stop");
+    }
+
+    public static void startupHook() {
+        out("Jfr startupHook() - start");
+        out(dumpJfrOptions(", "));
+        assert instance == null;
+        if (JfrOptions.getStartRecordingAutomatically()) {
+            try {
+                // JFR-TODO fix JfrOptions to parse configuration name instead of using default "default"
+                instance = new JfrAutoSessionManager();
+                String fn = (JfrOptions.getRecordingFileName() != null && JfrOptions.getRecordingFileName().length() > 0) ? JfrOptions.getRecordingFileName() : DEFAULT_JFR_FILENAME;
```

Comment:
Can the DEFAULT_FILENAME actually be generated to include time/date or something such that its unique across multiple runs of the native image?

Reply:
done.